packages/components/src/side_scroller/index.jsx

```diff
@@ -0,0 +1,136 @@
+import Controls from './controls';
+import { DISABLE_SCROLL_DISTANCE } from './constants';
+
+const SideScroller = ({ className: passedClassName, children, ...cleanProps }) => {
+  const [canScrollLeft, setCanScrollLeft] = useState(false);
+  const [canScrollRight, setCanScrollRight] = useState(false);
+  const [height, setHeight] = useState(0);
+
+  const containerRef = useRef(null);
+  const contentRef = useRef(null);
+  const containerWidthRef = useRef(null);
+  const contentWidthRef = useRef(null);
+  const animationIdRef = useRef(null);
+  const startXRef = useRef(null);
+  const startPositionRef = useRef(null);
+  const preventClickRef = useRef(null);
+  const stopListeningToResizeRef = useRef(null);
+
+  const className = clsx(styles.root, passedClassName);
+
+  // Fixes issue with most browsers reducing size of the scrollable element children
+  // as if to compensate for scrollbars, even when they are hidden
+  const style = { height };
+
+  const getScrollableNode = () => containerRef.current;
+
+  const stopScrollAnimation = () => {
+    if (animationIdRef.current) global.cancelAnimationFrame(animationIdRef.current);
+  };
+
+  const updateScroll = () => {
+    const { scrollLeft } = containerRef.current;
+    const maxScrollPosition = contentWidthRef.current - containerWidthRef.current;
+
+    const nextCanScrollLeft = scrollLeft > 0;
+    const nextCanScrollRight = scrollLeft < maxScrollPosition;
+
+    setCanScrollLeft(nextCanScrollLeft);
+    setCanScrollRight(nextCanScrollRight);
+  };
+
+  const updateSizes = () => {
+    const { width: containerWidth } = size(containerRef.current);
+    const { height: nextHeight } = size(contentRef.current);
+
+    containerWidthRef.current = containerWidth;
+    contentWidthRef.current = getScrollableNode().scrollWidth;
+
+    updateScroll();
+    setHeight(nextHeight);
+  };
+
+  const handleDragStart = (event) => {
+    // Prevents DOM nodes like images from being dragged
+    event.preventDefault();
+    startXRef.current = eventCoordinates(event, 'pageX').pageX;
+    startPositionRef.current = getScrollableNode().scrollLeft;
+    stopScrollAnimation();
+  };
+
+  const handleDrag = (event) => {
+    const diff = startXRef.current - eventCoordinates(event, 'pageX').pageX;
+    const shift = Math.abs(diff);
+    const newPosition = startPositionRef.current + diff;
+
+    if (shift >= DISABLE_SCROLL_DISTANCE) {
+      preventClickRef.current = true;
+
+      // Prevents vertical scroll on touch devices
+      event.preventDefault();
+    }
+
+    getScrollableNode().scrollLeft = newPosition;
+  };
+
+  const handleClickCapture = (event) => {
+    if (preventClickRef.current) {
+      stopEvent(event);
+      preventClickRef.current = false;
+    }
+  };
+
+  const handleAnimateScroll = (callback) => {
+    animationIdRef.current = global.requestAnimationFrame(callback);
+  };
+
+  useEffect(() => {
+    stopListeningToResizeRef.current = eventproxy('resize', updateSizes);
```

Comment:
We have to really watch out for deps here. Using functions inside `useEffect` without specifying them in `deps` can cause weird effects of state being outdated. I don't have a quick solution for refactoring this - would need to spend a bit more time on it.

@ozerovs do you get what I am saying? 


https://overreacted.io/a-complete-guide-to-useeffect/#but-i-cant-put-this-function-inside-an-effect

Reply:
I do get it. 
It seems that solution can be - wrapping all the functions that are being used in useCallback, and pass it as a dependency, no?

Reply:
I am torn a bit here. It currently works because none of those handlers use stuff from state but it will be very easy to introduce bugs.

Suggestion:
- Don't reuse handlers, extract functions. eg `handleScroll` and `handleLoad` (makes it a bit easier to understand and helps with tracking deps)
- All functions used inside `useEffect` need to have `useCallback` (you don't have to change the handlers!)
- Mention these functions you just covered with `useCallback` in effect deps